### DISCUSSION TITLE
Crash in Clip multiple rasters function because of CRS error

### DISCUSSION THREAD

**Discussion by Go-Tak** [semiautomaticgit/SemiAutomaticClassificationPlugin#197] [General]
win11(64)
QGIS:3.16.12-Hannover
python 3.7

When press Run button of Clip multiple raster,  I got the following error message.
#########
 AttributeError: 'NoneType' object has no attribute 'replace' 
Traceback (most recent call last):
  File "C:/Users/go_ta/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\SemiAutomaticClassificationPlugin\maininterface\clipmultiplerasters.py", line 82, in clipRastersAction
    self.clipRasters()
  File "C:/Users/go_ta/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\SemiAutomaticClassificationPlugin\maininterface\clipmultiplerasters.py", line 230, in clipRasters
    eCrs = cfg.utls.getCrsGDAL(l)
  File "C:/Users/go_ta/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\SemiAutomaticClassificationPlugin\core\utils.py", line 7360, in getCrsGDAL
    return crs.replace(' ', '')
AttributeError: 'NoneType' object has no attribute 'replace'
######### 

Then I tried to identify the place of error occurred by adding the codes in utils.py located under core directory.

########################################################################
# file validation

1. def getCrsGDAL(self, layerPath):
2. 		l = cfg.ogrSCP.Open(layerPath)
3. 		if l is None:
4. 			from osgeo import gdal, gdalconst
5. 			#l = cfg.gdalSCP.Open(layerPath, cfg.gdalSCP.GA_ReadOnly)
6. 			l = gdal.Open(layerPath, gdalconst.GA_ReadOnly)
7.                     if` l is None:
8.                        crs = None
9.                        f = open('c:/Users/go_ta/Downloads/tmp.txt', 'w')
10.                        f.write('crs is None:l is None')
11.                        f.close()
12.                  else:
13.                    try:            
14.             # check projections
15.                     crs = l.GetProjection()
16.                     f = open('c:/Users/go_ta/Downloads/tmp1.txt', 'w')
17.                     f.write('crs result:'+crs)
18.                     f.close()
19.                     if len(crs) == 0:
20.                        crs = None
21.                        f = open('c:/Users/go_ta/Downloads/tmp.txt', 'w')
22.                        f.writ e('crs is None'+l.GetProjection())
23.                        f.close()
24.           # in case of errors
25.                 except Exception as err:
26.                    crs = None
27.                    f = open('c:/Users/go_ta/Downloads/tmp.txt', 'w')
28.                    f.write('crs is None:exception:'+layerPath)
29.                    f.close()
30. 

######################################################################

Finally I found that error occurred in this statement `crs = l.GetProjection()`. Although this statement works fine in Jupyter.

I don't know why the above simple statement cause an error. 

Does someone know the way to solve this issue?

**Comment by semiautomaticgit** [comment]
Hello and thank you for reporting this issue.
Possibly , it could be related to this QGIS issue https://lists.osgeo.org/pipermail/qgis-developer/2021-November/064212.html
Possibly you could try to install a previous version of QGIS.
However, I'll look into this.

Also, please copy the log file following these steps https://semiautomaticclassificationmanual.readthedocs.io/en/latest/faq.html#how-can-i-report-an-error

**Reply by Go-Tak** [reply]
> Hello and thank you for reporting this issue. Possibly , it could be related to this QGIS issue https://lists.osgeo.org/pipermail/qgis-developer/2021-November/064212.html Possibly you could try to install a previous version of QGIS. However, I'll look into this.
> 
> Also, please copy the log file following these steps https://semiautomaticclassificationmanual.readthedocs.io/en/latest/faq.html#how-can-i-report-an-error

Thank you for the response.

I tried version 3.14 & 3.18 although both failed and the screen of Clip multiple rasters freezes.

I tried to get a log information as you indicated and attached below.  

I also attach the error message when executed l.GetProjection() without try -except trap.

When I see the Error.txt, it seems that there is different PROJ installed. Then I remember that I have 2 Python, one is Anaconda Python and the other is straight Python. I don't know how to set Python library used in QGIS and QGIS plugin. 

Although both python version is not 3.7.0 as recorded in error message, Anaconda Python is 3.7.10 and the other 3.7.6.


[log.txt](https://github.com/semiautomaticgit/SemiAutomaticClassificationPlugin/files/7480383/log.txt)
[Error.txt](https://github.com/semiautomaticgit/SemiAutomaticClassificationPlugin/files/7480728/Error.txt)

**Reply by Go-Tak** [reply]
Sorry to bother you.

I solved this error by setting environment variable PROJ_LIB to a value of %PROJ_LIB% variable.

**Comment by semiautomaticgit** [comment]
Thank you very much for your feedback.
I'm glad that you solved the error.
